delete cmake code

dundargoc opened the pull request.

dundargoc committed:
build(cmake): remove unnecessary *-prereqs targets

Running the tests on their own works just fine.

dundargoc committed:
build(cmake): simplify and speed up the uninstall target

More specifically, replace exec_program with file(REMOVE ...) so that
the uninstall target is run during the build stage instead of the
configure stage, significantly speeding up the target.

The code snippet that was removed is taken from the cmake FAQ
https://gitlab.kitware.com/cmake/community/-/wikis/FAQ#can-i-do-make-uninstall-with-cmake.
However, this uses undocumented features such as IMMEDIATE when calling
configure_file, which is an artifact from cmake 2.x (it's so old it's
difficult to find information on it). Similarly, this particular code
snippet has been around for a long time and originated from the cmake
mailing lists. Based on this I believe the in-file was a workaround for
the limitations of cmake back then and that it's not required anymore.

dundargoc committed:
build(cmake): remove unnecessary globbing with file

There's no need to use globbing since there's only one file.

dundargoc committed:
build(cmake): simplify def_cmd_target function

Instead of appending to a command output, append to an existing target
instead. The primary benefit is intermediary ...-cmd targets aren't
needed, we can instead append commands to the relevant target directly.

justinmk commented on CMakeLists.txt:
```diff
@@ -676,17 +681,6 @@ if(BUSTED_PRG)
   set(FUNCTIONALTEST_PREREQS nvim printenv-test printargs-test shell-test streams-test tty-test ${GENERATED_HELP_TAGS})
   set(BENCHMARK_PREREQS nvim tty-test)
 
-  # Useful for automated build systems, if they want to manually run the tests.
```
good call. The more random things we have, the harder it is to make sense of the system, document it, and know if these little hidden features are even working :)

justinmk commented on CMakeLists.txt:
```diff
@@ -618,29 +618,34 @@ find_program(SHELLCHECK_PRG shellcheck)
+    COMMAND ${STYLUA_PRG} --color=always --check runtime/)
 else()
-  add_custom_command(OUTPUT lintlua-cmd APPEND COMMAND ${CMAKE_COMMAND} -E echo "STYLUA_PRG not found")
+  add_custom_command(TARGET lintlua COMMAND ${CMAKE_COMMAND} -E echo "STYLUA_PRG not found")
```
makes sense, nice improvement

justinmk commented on CMakeLists.txt:
```diff
@@ -767,17 +760,8 @@ if(BUSTED_LUA_PRG)
-endif()
-
+add_custom_target(uninstall
+  COMMAND ${CMAKE_COMMAND} -P ${PROJECT_SOURCE_DIR}/cmake/UninstallHelper.cmake)
```
@Borwe was there a particular reason you chose to do this as a `*.in` file ? https://github.com/neovim/neovim/pull/14149 does not explain.

dundargoc replied:
This is taken straight from the [cmake FAQ](https://gitlab.kitware.com/cmake/community/-/wikis/FAQ#can-i-do-make-uninstall-with-cmake). This particular code snippet has existed for a VERY long time (since around 2010). The `IMMEDIATE` for `configure_file` is an undocumented keyword kept for backwards compatibility. It was officially "removed" in 2.x. It's so long ago that finding information about it is difficult. I think it's one of those "Well, no need to change it if it works" situations taken to the extreme.

justinmk replied:
perfect! would you mind including that in the commit message?

dundargoc replied:
Good idea.

justinmk reviewed (approved):
nice cleanup!

dundargoc committed:
build: remove FindLua.cmake since it's already built into cmake

FindLua.cmake is a copy from the cmake repo:
https://github.com/Kitware/CMake/blob/0419ecbcad7719614349a07189b45e341a8f2c69/Modules/FindLua.cmake.
It's a cmake module, meaning it's already shipped with cmake by default.
There have been two changes done to our version of FindLua.cmake.

The first change is that

include(${CMAKE_CURRENT_LIST_DIR}/FindPackageHandleStandardArgs.cmake)

was changed to

include(FindPackageHandleStandardArgs.cmake)

This change is required only because we have imported FindLua.cmake
module but not FindPackageHandleStandardArgs module. Had FindLua been
called as a module as intended then this file would not need changing.

The second change is that support for Lua 5.4 is added. However, support
for any version of Lua except for 5.1 is disabled since
https://github.com/neovim/neovim/pull/16633/commits/e322b5c864c771f774ddfea32f6cd5190a1af419.

Because these changes from the upstream FindLua.cmake is unnecessary I
believe we can and should use the builtin FindLua.cmake instead of our
own.
